pinot-transport/src/main/java/com/linkedin/pinot/requestHandler/MultipleOrEqualitiesToInClauseFilterQueryTreeOptimizer.java

```diff
@@ -0,0 +1,137 @@
+ * limitations under the License.
+ */
+
+package com.linkedin.pinot.requestHandler;
+
+import com.google.common.base.Splitter;
+import com.linkedin.pinot.common.request.FilterOperator;
+import com.linkedin.pinot.common.utils.StringUtil;
+import com.linkedin.pinot.common.utils.request.FilterQueryTree;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.TreeSet;
+
+
+/**
+ * Optimizer that collapses multiple OR clauses to IN clauses. For example, <code>a = 1 OR a = 2 OR a =
+ * 3</code> gets turned to <code>a IN (1, 2, 3)</code>.
+ */
+public class MultipleOrEqualitiesToInClauseFilterQueryTreeOptimizer implements FilterQueryTreeOptimizer {
+  @Override
+  public FilterQueryTree optimize(FilterQueryTree filterQueryTree) {
+    return optimize(filterQueryTree, null);
+  }
+
+  private FilterQueryTree optimize(FilterQueryTree filterQueryTree, FilterQueryTree parent) {
+    if (filterQueryTree.getOperator() == FilterOperator.OR) {
+      Map<String, Set<String>> columnToValues = new HashMap<>();
+      List<FilterQueryTree> nonEqualityOperators = new ArrayList<>();
+
+      // Collect all equality/in values and non-equality operators
+      for (FilterQueryTree childQueryTree : filterQueryTree.getChildren()) {
+        if (childQueryTree.getOperator() == FilterOperator.EQUALITY || childQueryTree.getOperator() == FilterOperator.IN) {
+          List<String> childValues = valueDoubleTabListToElements(childQueryTree.getValue());
+
+          if (!columnToValues.containsKey(childQueryTree.getColumn())) {
+            columnToValues.put(childQueryTree.getColumn(), new TreeSet<String>(childValues));
+          } else {
+            columnToValues.get(childQueryTree.getColumn()).addAll(childValues);
+          }
+        } else {
+          nonEqualityOperators.add(childQueryTree);
+        }
+      }
+
+      // If we have at least one column to return
+      if (!columnToValues.isEmpty()) {
+        // We can eliminate the OR node if there is only one column with multiple values
+        if (columnToValues.size() == 1 && nonEqualityOperators.isEmpty()) {
+          Map.Entry<String, Set<String>> columnAndValues = columnToValues.entrySet().iterator().next();
+
+          return buildFilterQueryTreeForColumnAndValues(columnAndValues);
+        }
+
+        // Rebuild the predicates
+        ArrayList<FilterQueryTree> newChildren = new ArrayList<>();
+        for (Map.Entry<String, Set<String>> columnAndValues : columnToValues.entrySet()) {
+          newChildren.add(buildFilterQueryTreeForColumnAndValues(columnAndValues));
+        }
+        newChildren.addAll(nonEqualityOperators);
+        return new FilterQueryTree(null, null, FilterOperator.OR, newChildren);
+      }
+    } else if (filterQueryTree.getChildren() != null){
```

Comment:
The children list is null to avoid allocating an empty array list for all leaf nodes of the filter tree.